gh-103509: PEP 697 -- Limited C API for Extending Opaque Types

encukou opened the pull request:
[PEP 697](https://peps.python.org/pep-0697/) describes the change. Some details it doesn't cover:

- `_PyHeapType_GET_MEMBERS` is moved from `Include/internal` to the only `.c` file that uses it.
- Tests demonstrate the API is usable for the purpose, and verify the layout with brutish-force.

TODO:
- [x] What's New entry
- [x] Fix the `alignof` & `max_align_t`

<!-- gh-issue-number: gh-103509 -->
* Issue: gh-103509
<!-- /gh-issue-number -->

encukou committed:
gh-103509: PEP 697 -- Limited C API for Extending Opaque Types

encukou committed:
Add a blurb

encukou committed:
Fix PyMember_SetOne error return (thanks MSVC!)

encukou committed:
Add What's New entry

encukou committed:
Fix warning in tests

encukou committed:
Work around lack of alignof & max_align_t

encukou committed:
Use the same Sphinx role for Py_TPFLAGS_ITEMS_AT_END as for other type flags

Now is not the time to fix that.

encukou committed:
Add ALIGNOF_MAX_ALIGN_T to configure & PC/pyconfig.h

encukou committed:
Use ALIGNOF_MAX_ALIGN_T

encukou committed:
Fix typo

encukou committed:
Don't include <stdalign.h>, it's not always available on Windows

encukou committed:
Define ALIGNOF_MAX_ALIGN_T with long double if it's not available

encukou committed:
tests: Compute data_offset in C to avoid overflow issues

arhadthedev commented on Modules/_testcapi/heaptype_relative.c:
```diff
@@ -0,0 +1,289 @@
+
+    result = Py_BuildValue("OOOKnn", base, sub, instance,
+                           (unsigned long long)data_ptr,
+                           (Py_ssize_t)(data_ptr - (void*)instance),
```
```suggestion
                           (Py_ssize_t)(data_ptr - (char*)instance),
```

arhadthedev commented on Modules/_testcapi/heaptype_relative.c:
```diff
@@ -0,0 +1,289 @@
+    if (!instance) {
+        goto finally;
+    }
+    void *data_ptr = PyObject_GetTypeData(instance, (PyTypeObject *)sub);
```
```suggestion
    char *data_ptr = PyObject_GetTypeData(instance, (PyTypeObject *)sub);
```

arhadthedev reviewed (changes requested):
`void*` cannot participate in address arithmetics. You need to cast both `data_ptr` and `instance` to `char *` to get distance in bytes.

arhadthedev committed:
Merge branch 'main' into extend-opaque-sq

encukou committed:
Cast to `char*` for pointer arithmetic

Co-authored-by: Oleg Iarygin <dralife@yandex.ru>

encukou committed:
Fix ALIGNOF_MAX_ALIGN_T value for 32-bit Windows

(I set it low on purpose, so the CI tests would find the issue.)

encukou committed:
Fix C++ behaviour and comments for ALIGNOF_MAX_ALIGN_T

encukou committed:
Merge branch 'main' into extend-opaque-sq

encukou committed:
Don't rely on PyObject* being aligned to ALIGNOF_MAX_ALIGN_T

Smaller objects might not be aligned to ALIGNOF_MAX_ALIGN_T.
The offsets calculated for PEP 697 should be aligned, though.

encukou committed:
Merge in main branch

erlend-aasland commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
 
    .. versionadded:: 3.10
+
+.. c:function:: PyObject* PyObject_GetTypeData(PyObject *o, PyTypeObject *cls)
```
```suggestion
.. c:function:: void *PyObject_GetTypeData(PyObject *o, PyTypeObject *cls)
```

erlend-aasland commented on Doc/c-api/type.rst:
```diff
@@ -331,25 +331,57 @@ The following functions and structs are used to create
+        directly.
+      - With the :const:`Py_TPFLAGS_ITEMS_AT_END` flag.
+
+   .. c:member:: int flags
```
```suggestion
   .. c:member:: unsigned int flags
```

erlend-aasland replied:
(BTW, some of these clean-ups could be backported to the 3.11 docs)

encukou replied:
They could!
Alas, I don't think it'll be enough of a priority for *me* to get to it.

erlend-aasland replied:
I'll see if I can find time to do that.

erlend-aasland commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
+
+   Get a pointer to subclass-specific data reserved for *cls*.
+
+   The object *o* **must** be an instance of *cls*, and *cls* must have been
```
Observation: only one of the _must_'s are emphasised.

encukou replied:
```suggestion
   The object *o* must be an instance of *cls*, and *cls* must have been
```

erlend-aasland replied:
Yeah, that would work.

erlend-aasland commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
+   negative :c:member:`PyType_Spec.basicsize`.
+   Python does not check this.
+
+   On error, set an exception and return a negative value.
```
AFAICS, this function always succeeds; an exception is never set.

Return value is always >= 0.

encukou replied:
The current implementation might not raise exceptions, but they may be added in the future. Users must implement error handling.

erlend-aasland replied:
That seems like a weird decision to me; if the current implementation does not raise exceptions, surely the current docs must reflect that fact. If the implementation changes, we change the docs. No?

erlend-aasland replied:
Expanding: the runtime will segfault if the user does not _set an exception_ if any of these APIs indicate error. It is an error to return `NULL` from a function whose return value is `PyObject *` without an exception set.

erlend-aasland replied:
```suggestion
   This function always succeeds.
```

erlend-aasland replied:
Wait, I think I see where you are heading. Yeah, you're good to go. (I just needed another cup of coffee to get this straight.)

erlend-aasland commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
+   Get a pointer to per-item data for a class with
+   :const:`Py_TPFLAGS_ITEMS_AT_END`.
+
+   On error, set an exception and return ``NULL``.
+   :py:exc:`TypeError` is raised if *o* does not have
+   :const:`Py_TPFLAGS_ITEMS_AT_END` set.
```
AFAICS, this function always succeeds; an exception is never set.

`TypeError` is not raised if `Py_TPFLAGS_ITEMS_AT_END` is not set; there's only an debug assertion.

encukou replied:
Good catch, thanks!

erlend-aasland commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
+   created using negative :c:member:`PyType_Spec.basicsize`.
+   Python does not check this.
+
+   On error, set an exception and return ``NULL``.
```
Again, no exception is set AFAICS. There is no error checking apart from a debug assert.

encukou replied:
The current implementation might not raise exceptions, but they may be added in the future. Users must implement error handling.

erlend-aasland commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
+
+   .. versionadded:: 3.12
+
+.. c:function:: PyObject* PyObject_GetItemData(PyObject *o)
```
```suggestion
.. c:function:: void *PyObject_GetItemData(PyObject *o)
```

erlend-aasland commented on Misc/NEWS.d/next/C API/2023-04-13-16-54-00.gh-issue-103509.A26Qu8.rst:
```diff
@@ -0,0 +1,5 @@
+Added C API for extending types whose instance memory layout is opaque:
+:c:member:`PyType_Spec.basicsize` can now be zero or negative,
+:c:func:`PyObject_GetTypeData` can be used to get subclass-specific data,
+and :c:macro:`Py_TPFLAGS_ITEMS_AT_END` can be used to safely extend
+variable-size objects. See :pep:`697` for details.
```
The NEWS entry should also mention `PyType_GetTypeDataSize`. Also, I think the change in semantics for `PyType_Spec.basicsize` should be better communicated.

The Limited API additions should also be mentioned.

encukou replied:
IMO, a NEWS blurb should be brief. It doesn't need to document the changes precisely, especially if it links to a detailed change document.

erlend-aasland commented on Misc/stable_abi.toml:
```diff
@@ -2397,3 +2397,12 @@
 [const.Py_AUDIT_READ]
     added = '3.12'  # Before 3.12, available in "structmember.h"
+
+[function.PyObject_GetTypeData]
+    added = '3.12'
+[function.PyType_GetTypeDataSize]
+    added = '3.12'
```
Missing an entry for `PyObject_GetItemData`?

encukou replied:
That's [intentionally not added to the limited API](https://peps.python.org/pep-0697/#list-of-new-api).
(I don't understand the minute details of variable-sized objects enough to feel safe adding it.)

erlend-aasland commented on Lib/test/test_capi/test_misc.py:
Would you mind wrapping the "heap type relative" tests in a `unittest.TestCase` subclass?

encukou replied:
OK.
It'll not be not ideal -- hopefully in the future other tests will be mingled with this way of defining types -- but I guess it'll be better than making the current monster of a class bigger.

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
+    sizeof(PyObject*),
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC
+        | Py_TPFLAGS_ITEMS_AT_END,
+    HeapCCollection_slots
```
```suggestion
    .name = "_testcapi.HeapCCollection",
    .basicsize = sizeof(PyVarObject),
    .itemsize = sizeof(PyObject*),
    .flags = (Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE |
              Py_TPFLAGS_HAVE_GC | Py_TPFLAGS_ITEMS_AT_END),
    .slots = HeapCCollection_slots,
```

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
+    {Py_tp_traverse, HeapCCollection_traverse},
+    {Py_tp_clear, HeapCCollection_clear},
+    {Py_tp_dealloc, HeapCCollection_dealloc},
+    {Py_tp_doc, (char*)HeapCCollection_doc},
```
```suggestion
    {Py_tp_doc, (void *)HeapCCollection_doc},
```

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
+}
+
+static int
+HeapCCollection_clear(PyObject *self) {
```
```suggestion
HeapCCollection_clear(PyObject *self)
{
```

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
+    return 0;
+}
+
+static void HeapCCollection_dealloc(PyObject *self) {
```
```suggestion
static void
HeapCCollection_dealloc(PyObject *self)
{
```

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -1096,5 +1200,12 @@ _PyTestCapi_Init_Heaptype(PyObject *m) {
+    if (HeapCCollection == NULL) {
+        return -1;
+    }
+    PyModule_AddObject(m, "HeapCCollection", HeapCCollection);
```
```suggestion
    if (PyModule_AddObject(m, "HeapCCollection", HeapCCollection) < 0) {
        Py_DECREF(HeapCCollection);
        return -1;
    }
```

...or:

```suggestion
    int rc = PyModule_AddType(m, (PyTypeObject *)HeapCCollection);
    Py_DECREF(HeapCCollection);
    if (rc < 0) {
        return -1;
    }
```

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
+        goto finally;
+    }
+    PyObject **data = PyObject_GetItemData(self);
+    if (!data) {
+        goto finally;
+    }
```
You need to set an exception before erring out.

encukou replied:
`PyObject_GetItemData` sets an exception on error.

erlend-aasland replied:
Um, I think I misreading something, then. Exactly where is it setting an exception?

encukou replied:
Like most other C API lunctions, whenever `PyObject_GetItemData` returns NULL, it sets an exception. 
If it *currently* has no error cases, and won't actually return `NULL`, that's an internal implementation detail. The tests show how a user should use the API: you should check for errors and propagate exceptions.

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
+        return PyErr_Format(PyExc_IndexError, "index %zd out of range", i);
+    }
+    PyObject **data = PyObject_GetItemData(self);
+    if (!data) {
+        return NULL;
+    }
```
You need to set an exception before erring out.

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
+HeapCCollection_traverse(PyObject *self, visitproc visit, void *arg)
+{
+    PyObject **data = PyObject_GetItemData(self);
+    if (!data) {
+        return -1;
+    }
```
You need to set an exception before erring out.

erlend-aasland commented on Modules/_testcapi/heaptype.c:
```diff
@@ -973,6 +973,110 @@ static PyType_Spec HeapCTypeSetattr_spec = {
+static int
+HeapCCollection_clear(PyObject *self) {
+    PyObject **data = PyObject_GetItemData(self);
+    if (!data) {
+        return -1;
+    }
```
You need to set an exception before erring out.

erlend-aasland commented on Modules/_testcapi/heaptype_relative.c:
```diff
@@ -0,0 +1,289 @@
+    if (!instance) {
+        goto finally;
+    }
+    char *data_ptr = PyObject_GetTypeData(instance, (PyTypeObject *)sub);
+    if (!data_ptr) {
+        goto finally;
+    }
```
You need to set an exception before erring out.

erlend-aasland replied:
Ditto for the usages in the rest of the file.

erlend-aasland commented on Modules/_testcapi/heaptype_relative.c:
```diff
@@ -0,0 +1,289 @@
+        goto finally;
+    }
+    Py_ssize_t data_size = PyType_GetTypeDataSize((PyTypeObject *)sub);
+    if (data_size < 0) {
+        goto finally;
+    }
```
`PyType_GetTypeDataSize` cannot return negative values; the check is not needed.

erlend-aasland replied:
Ditto for the usages in the rest of the file.

encukou replied:
Here I want to test with intended usage, rather than rely on current details.

erlend-aasland replied:
Fair enough

erlend-aasland commented on Modules/_testcapi/heaptype_relative.c:
```diff
@@ -0,0 +1,289 @@
+        return -1;
+    }
+
+    PyModule_AddIntConstant(m, "alignof_max_align_t", ALIGNOF_MAX_ALIGN_T);
```
```suggestion
    if (PyModule_AddIntConstant(m, "alignof_max_align_t", ALIGNOF_MAX_ALIGN_T) < 0) {
        return -1;
    }
```

erlend-aasland replied:
Or even better, rewrite to use `PyModule_AddIntMacro` (need to change the .py test as well).

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3556,6 +3563,14 @@ static const PySlot_Offset pyslot_offsets[] = {
+ * (like _Py_ALIGN_UP, but for a size rather than pointer)
+ */
+static Py_ssize_t
+_align_up(Py_ssize_t size) {
```
```suggestion
_align_up(Py_ssize_t size)
{
```

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3695,6 +3710,20 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
                     vectorcalloffset = memb->offset;
                 }
+                if (memb->flags & Py_RELATIVE_OFFSET) {
+                    if(spec->basicsize > 0) {
```
```suggestion
                    if (spec->basicsize > 0) {
```

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3695,6 +3710,20 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
+                            "With Py_RELATIVE_OFFSET, basicsize must be negative.");
+                        goto finally;
+                    }
+                    if(memb->offset < 0 || memb->offset >= -spec->basicsize) {
```
```suggestion
                    if (memb->offset < 0 || memb->offset >= -spec->basicsize) {
```

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3889,6 +3944,15 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
                 memcpy(_PyHeapType_GET_MEMBERS(res), slot->pfunc, len);
                 type->tp_members = _PyHeapType_GET_MEMBERS(res);
+                PyMemberDef *memb;
+                unsigned i;
```
```suggestion
                Py_ssize_t i;
```

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3889,6 +3944,15 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
+                PyMemberDef *memb;
+                unsigned i;
+                for (memb = _PyHeapType_GET_MEMBERS(res), i = nmembers;
+                        i > 0; ++memb, --i) {
```
```suggestion
                for (memb = _PyHeapType_GET_MEMBERS(res), i = nmembers;
                     i > 0; ++memb, --i)
                {
```

erlend-aasland commented on PC/python3dll.c:
Also `PyObject_GetItemData`?

erlend-aasland commented on Python/structmember.c:
FTR, there's no test coverage for these added branches.

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3695,6 +3710,20 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
                     assert(memb->flags == READONLY);
                     vectorcalloffset = memb->offset;
                 }
+                if (memb->flags & Py_RELATIVE_OFFSET) {
```
There's no test coverage for anything between line 3713 and 3726.

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3824,6 +3853,32 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
+        /* Inheriting variable-sized types is limited */
+        if (base->tp_itemsize
+            && !((base->tp_flags | spec->flags) & Py_TPFLAGS_ITEMS_AT_END))
+        {
+            PyErr_SetString(
+                PyExc_SystemError,
+                "Cannot extend variable-size class without Py_TPFLAGS_ITEMS_AT_END.");
+            goto finally;
+        }
```
There's no test coverage for this branch.

erlend-aasland reviewed (commented):
Very nice!

Sorry for the late review; it took some time to dig through everything.

encukou committed:
Raise TypeError on missing flag

encukou committed:
Wrap new tests in a class

encukou committed:
Use PyModule_AddIntMacro in tests

encukou committed:
Apply suggestions from code review

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

encukou committed:
Test failure of extending non-ITEMS_AT_END variable-sized types

encukou committed:
Test error cases around members

encukou committed:
Merge in the main branch

arhadthedev commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
+
+.. c:function:: void *PyObject_GetTypeData(PyObject *o, PyTypeObject *cls)
+
+   Get a pointer to subclass-specific data reserved for *cls*.
```
For a new extension developer it may seem like getting class-specific metadata. Probably, something like this (but less clumsy) would be better:

```suggestion
   Get a pointer to a part of instance memory space of *o* reserved for *cls* level of an inheritance chain.
```

encukou replied:
Hm. Would this work?
```suggestion
   Get a pointer to the memory space of *o* reserved for data specific to subclass *cls*.
```

arhadthedev replied:
> Hm. Would this work?

Looks good.

arhadthedev commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
+
+.. c:function:: Py_ssize_t PyType_GetTypeDataSize(PyTypeObject *cls)
+
+   Return the size of the memory reserved for *cls*, i.e. the size of the
```
The same suggestion as above.

```suggestion
   Return the size of the instance memory space reserved for *cls*, i.e. the size of the
```

arhadthedev commented on Doc/c-api/object.rst:
```diff
@@ -395,3 +395,42 @@ Object Protocol
+
+.. c:function:: void *PyObject_GetItemData(PyObject *o)
+
+   Get a pointer to per-item data for a class with
```
The same suggestion as above.

```suggestion
   Get a pointer to variable-length instance data not belonging to any class (level) in the inheritance chain.
```

encukou replied:
This PR adds the concept of data being reserved for particular (sub)classes/levels in the chain. 
`PyObject_GetItemData` should work with existing classes, where no data is tied to a subclass.
Also, variable-sized data *never* belongs to a specific subclass, even with this PR.

So IMO the original is better. It can probably still be improved though.

arhadthedev commented on Doc/c-api/type.rst:
```diff
@@ -331,25 +331,57 @@ The following functions and structs are used to create
+
+   .. c:member:: int itemsize
+
+      Size of one element of a variable-size type, in bytes
```
A forgotten full stop.

```suggestion
      Size of one element of a variable-size type, in bytes.
```

arhadthedev reviewed (commented):
*Please, keep in mind that I neither participated in the PR discussion nor seen the PEP before. So I looked at the rendered docs in the same fashion as an outside user would do.*

Here are three optional rough proposals and a forgotten full stop. However, they are non-binding suggestions that can be properly considered later, after 3.12b1 is out (otherwise the two remaining weeks will pass swiftly in the bikeshedding).

I've limited myself to the documentation because I have no expertise in the C API to check such a volume of code in such a short period of time.

encukou committed:
Apply suggestions from code review

Co-authored-by: Oleg Iarygin <oleg@arhadthedev.net>

encukou reviewed (commented):
Thank you! Fresh eyes is exactly what the documentation needs.

It will also need a tutorial to bring the concepts together -- but when trying to write a tutorial, I keep finding rough edges that should be fixed first :)

encukou committed:
Merge in the main branch
